### ISSUE TITLE
Fast Clustering Does Not Find All Clusters on the First Attempt - Running Un-Clustered Results Back Results in More Clusters

### ISSUE THREAD

**Issue by searchsolved** [UKPLab/sentence-transformers#1408]
I'm using this script as a base for a project https://github.com/UKPLab/sentence-transformers/blob/master/examples/applications/clustering/fast_clustering.py 

When I set  the threshold value high, I get tighter cluster groups as expected, which in turn means an increase in sentences which were not clustered. 

I've noticed if I run the sentence which were not clustered through a second time on their own, around 50% are clustered again. 

My question: Is this by design? Or should I look to running a second / third pass on items which were un-clustered the first time, or is this human error on my part?

**Comment by nreimers** [comment]
Yes, this can happen. It was designed to find quickly large communities in the vector space, not to assign all points to a cluster / trying to cluster all points.

You could do a second or third pass. Otherwise HDBSCAN is also a good algorithm. For larger number of points it is sadly a bit slow, but the it assigns more points to cluster. So HDBSCAN would be a good choice here.

The fast clustering algorithm is more designed when you have like 100k data points. Here missing some (smaller) clusters is usually not an issue. The missed clusters in the first pass are between two or more clusters.

**Comment by searchsolved** [comment]
Thanks for the detailed reply. I'll take a look at HDBSCAN. Appreciate it.
